CMSgov/mdct-mcr
# Cypress login

## Description
Modifies our custom 'authentication' module to take advantage of [Cypress sessions](https://docs.cypress.io/api/commands/session). Sessions allow us to reuse the auth tokens from a given login, ensuring that we don't need to relogin for every test that does not explicitly want to test the login flow. Sessions only persist per `.spec.` file so a future enhancement is making these tokens count for all tests, not just per file. That said, this change alone reduces our cypress runtime by up to a minute.

BEFORE
![Screen Shot 2022-11-04 at 2 06 10 PM](https://user-images.githubusercontent.com/57802560/200359435-ed2aed09-2990-4fda-b822-4b63690b2b2c.png)

AFTER
![Screen Shot 2022-11-07 at 11 07 12 AM](https://user-images.githubusercontent.com/57802560/200359467-ba30c65b-ef7c-4bd7-bc25-2159986d8c78.png)

These two config changes help us take advantage of this feature. You can also view [this guide on sessions](https://www.cypress.io/blog/2021/08/04/authenticate-faster-in-tests-cy-session-command/) for a brief idea of what we went for.
```
experimentalSessionAndOrigin: true,
 testIsolation: "off",
```

To further take advantage of these features, we upgraded Cypress to version 10. See the migration guide and GitHub actions updates below.
[Cypress GitHub Action releases](https://github.com/cypress-io/github-action/releases) (v4.0 adds Cypress 10 support)
[Cypress v10 migration guide](https://docs.cypress.io/guides/references/migration-guide#Config-Option-Changes)
Key changes:
- `config.json` -> `cypress.config.js`
- ~`plugins/index.js`~ -> `cypress.config.js` (within the config for e2e tests)


Other notes on changes:
1. `cy.visit("/")` was moved to 'afterEach' statements because our authentication will land on the homepage. When we reuse a session, we need to start back at the same spot because the session will not rerun that step.
2. `cy.waitUntil` is a neat community package we picked up because it allows us to only save the session once our tokens are ready. (doing a manual while loop with `cy.wait` caused the wait to fire as many times as possible regardless of token readiness. Hence why the community made this package)

### How to test
_See changed dependencies below_

Run cypress tests locally and ensure they pass! (they pass in GitHub actions!)

### Changed Dependencies
I suggest removing your `node_modules` folder from `tests/cypress` and running `yarn` in that folder to reinstall with cypress 10

## Code author checklist

## Reviewer checklist (two different people)
